PLAT-533 + PLAT-485: bugfix outgoing token resend

sexyhamster opened the pull request:
**- What I did**
- proper token for outgoing resending
- test that verifies proper token

ElSacramento commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -1704,3 +1707,192 @@ func TestVirtual_CallContractTwoTimes(t *testing.T) {
+
+			if callCount == 0 {
+				assert.Equal(t, payload.CallDelegationToken{}, request.DelegationSpec)
+				server.IncrementPulseAndWaitIdle(ctx)
```
increment must be before VCallResult response 
`пульс меняется до того как пришел VCallResult от контракта B `

sexyhamster replied:
well, pulse changes after VCallRequest sent and before VCallResult received.

ElSacramento commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -1704,3 +1707,192 @@ func TestVirtual_CallContractTwoTimes(t *testing.T) {
+
+		runnerMock.AddExecutionMock("Bar").AddStart(
+			func(ctx execution.Context) {
+				t.Log("ExecutionStart [A.Bar]")
```
remove t.log or change it  
https://insolar.atlassian.net/wiki/spaces/DEV/pages/1431568423/Core+v2+-+Logger+in+tests

sexyhamster replied:
switched to initialized logger from test context

ruz commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -661,7 +672,7 @@ func TestVirtual_Call_UnorderedMethod_From_UnorderedMethod(t *testing.T) {
 		func(ctx execution.Context) {
-			t.Log("ExecutionStart [A.Bar]")
+			logger.Debug("ExecutionStart [A.Bar]")
 		},
```
come on, do this kind of cleanups via NOISSUE branches

ruz commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -911,7 +926,7 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
 
 		runnerMock.AddExecutionMock("Foo").AddStart(
```
I think that it is better to drop mock of executing B. Just don't resend VCallRequest, check token in the message and generate manually VCallResponse.

ruz commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -1704,3 +1726,194 @@ func TestVirtual_CallContractTwoTimes(t *testing.T) {
+			SetInterference(barIsolation.Interference).
+			SetIsolation(barIsolation.State)
+		objectBExecutionMock := runnerMock.AddExecutionMock("Foo")
+		objectBExecutionMock.AddStart(
```
I think that it is better to drop mock of executing B. Just don't resend VCallRequest, check token in the message and generate manually VCallResponse.

sexyhamster replied:
not sure about this approach, It'll require more work, if something around VCallResponse will change.

ruz replied:
What if interface of execute adapter would change?

How this B's part is relevant to delegation token sent by call from A to B? As far as I know in this test the token is not checked by recipient. So it's irrelevant part. As I said many times messages are our boundaries for integration tests.

sexyhamster replied:
switched to message instead of execution. Probably we need to rethink most of the tests in this file. Most of them actually execute more than one request.

ruz commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -1704,3 +1726,194 @@ func TestVirtual_CallContractTwoTimes(t *testing.T) {
+			}
+
+			service := authentication.NewService(ctx, server.JetCoordinatorMock)
+			origToken := service.GetCallDelegationToken(
```
No-no, you don't use platform code for checks. You should check every field using test constants.

sexyhamster replied:
could you explain why? Result of GetCallDelegationToken output should be tested elsewhere. If GetCallDelegationToken logic will change, why do you want this test to fail?

ruz commented on ledger-core/virtual/integration/method_test.go:
```diff
@@ -1704,3 +1726,194 @@ func TestVirtual_CallContractTwoTimes(t *testing.T) {
+			assert.Equal(t, payload.CTMethod, request.CallType)
+			assert.Equal(t, uint32(1), request.CallSequence)
+			assert.Equal(t, outgoingCallRef, request.CallReason)
+			assert.Equal(t, payload.BuildCallFlags(barIsolation.Interference, barIsolation.State), request.CallFlags)
```
we test code around obtaining delegation token and attaching it to external calls. it's very strange that so many fields of VCallRequest are checked, but **ZERO** fields of VDelegatedCallRequest, VDelegatedCallResponse are checked. Also, VDelegatedRequestFinished is not checked for presence of the token.

sexyhamster replied:
This test tests token that is attached to outgoing call from contract. VDelegatedCallRequest and VDelegatedCallResponse should be tested elsewhere with their own tests. From my point of view, testing all the messages in all the tests is a waste of time.

sexyhamster commented:
I strongly disagree, you want this test to emulate a lot of the system logic and test not only this case but also bunch of other stuff. Other suff (VStateReport/VDelegatedCallRequest/VDelegatedCallResponse) should have their own test cases. The thing with naturally - add another test case and create a ticket for that.

sexyhamster commented:
OK, as discussed - I removed all the parts of the system that is outside of test scope. 
Double pulse is discribed in different case - C5205, not sure that this could be done via table test
Added one more case for execution that lasts 2 pulses.

ElSacramento commented on ledger-core/virtual/integration/outgoing_test.go:
```diff
@@ -0,0 +1,221 @@
+			CallMethod(objectBGlobal, classB, "Bar", []byte("123")).
+			SetInterference(barIsolation.Interference).
+			SetIsolation(barIsolation.State)
+		objectBExecutionMock := runnerMock.AddExecutionMock("Foo")
```
objectAExecutionMock

ElSacramento commented on ledger-core/virtual/integration/outgoing_test.go:
```diff
@@ -0,0 +1,221 @@
+
+			assert.Equal(t, firstPulse, request.CallOutgoing.Pulse())
+			if callCount == 0 {
+				assert.Equal(t, payload.CallDelegationToken{}, request.DelegationSpec)
```
you can use `request.DelegationSpec.IsZero()` (just proposal)

sexyhamster replied:
when it fails - here it will tell you that expectations is empty CallDelegationToken, but received CallDelegationToken with values. assert.True will assert that wanted true, got false, which is less verbose

sexyhamster replied:
well, on the second thought, I could switch to assert.Zero

ElSacramento commented on ledger-core/virtual/integration/outgoing_test.go:
```diff
@@ -0,0 +1,221 @@
+				ReturnArguments: []byte("finish B.Bar"),
+			})
+
+			return false // resend
```
wrong comment

sexyhamster replied:
fixed

ElSacramento commented on ledger-core/virtual/integration/outgoing_test.go:
```diff
@@ -0,0 +1,221 @@
+		typedChecker.VDelegatedCallRequest.Set(func(request *payload.VDelegatedCallRequest) bool {
+			assert.Zero(t, request.DelegationSpec)
+			assert.Equal(t, objectAGlobal, request.Callee)
+			assert.Equal(t, payload.BuildCallFlags(fooIsolation.Interference, fooIsolation.State), request.CallFlags)
```
why do you check flags? I do think this test isn't about isolation.

sexyhamster replied:
removed extra checks.

ElSacramento commented on ledger-core/virtual/integration/outgoing_test.go:
```diff
@@ -0,0 +1,195 @@
+				return false
+			}
+
+			assert.Equal(t, expectedToken, request.DelegationSpec)
```
check repeat flag, please
`Repeated Call`

sexyhamster replied:
done

ElSacramento commented on ledger-core/virtual/integration/outgoing_test.go:
```diff
@@ -0,0 +1,195 @@
+	assert.Equal(t, 2, typedChecker.VCallRequest.Count())
+	assert.Equal(t, 1, typedChecker.VCallResult.Count())
+	assert.Equal(t, 1, typedChecker.VStateReport.Count())
+	assert.Equal(t, 1, typedChecker.VDelegatedCallRequest.Count())
```
forget about VDelegatedRequestFinished
